keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -243,147 +233,162 @@ A `Bucket` is not deleted if it is bound to a `BucketRequest`.
-1. `bucketRequestName`: (required) the name of the `BucketRequest` associated with this access request. From the bucket request, COSI knows the `Bucket` instance and thus the backend bucket and its properties.
 1. `bucketAccessClassName`: (required) name of the `BucketAccessClass` specifying the desired set of policy actions to be set for a user identity or ServiceAccount.
+1. `serviceAccountName`: (optional) the name of a Kubernetes ServiceAccount, in the same namespace, which is expected to be linked to an identity in the cloud provider, such that a pod using the ServiceAccount is authenticated as the linked identity. In this case, COSI applies the policy defined in the `BAC.policyActionsConfigMap` to the identity. If `serviceAccountName` is omitted, a "minted" user is generated by the driver and stored as `BucketAccess.accountID`.
+1. `bucketRequestName`: (optional) the name of the `BucketRequest` associated with this access request within the same namespace as the BAR. From the BR, COSI knows the `Bucket` instance and thus the backend bucket and its properties. `bucketRequestName` and `BucketName` are mutually exclusive.
+1. `bucketName`: (optional) the name of the `Bucket` instance representing the backend bucket. `bucketRequestName` and `BucketName` are mutually exclusive.
 1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance. Set by COSI.
 1. `accessGranted`: if true access has been granted to the bucket. If false then access to the bucket has not been granted.
 
 #### BucketAccess
 
-A cluster-scoped administrative custom resource which encapsulates fields from the `BucketAccessRequest` (BAR) and the `BucketAccessClass` (BAC).  The purpose of the `BucketAccess` (BA) is to serve as an access communication path between provisioners and the central COSI controller.  The COSI controller creates a `BucketAccess` instance for a new `BucketAccessRequest`, and there is a 1:1 mapping between `BucketAccess` and `BucketAccessRequest`.
+A cluster-scoped, immutable, administrative custom resource which encapsulates fields from the [BucketAccessRequest](#bucketaccessrequest) (BAR) and the [BucketAccessClass](#bucketaccessclass) (BAC), and references a [Bucket](#bucket) instance.  The `BucketAccess` (BA) is an abstraction of a single access point to the backend bucket, and is instantiated when a new BAR is created.
+
+There is a 1:1 mapping between a BA and a BAR. Many workload pods can point to the same BA (via their BAR reference). There is a many:1 mapping of BAs the same `Bucket` instance.
```

Comment:
If a BA can not exist without BAR, then why does BA exist in the first place? Couldn't we simply merge BA and BAR into a single resource?

Reply:
To give some context:

- PV exists, because a PV can exist without PVC. Either because it was statically provisioned, or because you want to retain the state associated to it, i.e. the volume's content.
- Bucket exists, because a Bucket can exist without BucketRequest. Either because it was statically provisioned, or because you want to retain the state associated to it, i.e. the bucket's content.

But a BA can neither be created without BAR nor is there the need to retain it, as there's no state associated to a BA.

Reply:
I believe there is a _driverless_ case where an admin will create the BA manually and the user's BAR will reference this BA. In this case it seems that a BAC is not needed.
There is also the desire to not have the user specify access policy in their BAR. Instead, access is defined in the BAC and copied to the BA.

Reply:
Hold on. There is no driverless usecase. If you want to do things without a driver, then why use COSI

Reply:
> user can still reuse access policies by reusing the policy ConfigMap

The user may not have access to the policy config map. Also, in the last COSI meeting, there was discussion about removing the CM and instead the BAC would embed the access policy, similar to what is seen in the BA.

Reply:
The user also doesn't have access to the BAC. If you want to configure the access policy via a custom resource (instead of ConfigMap), then I think that resource should be namespaced, and it should probably have a different name, e.g. BucketAccessPolicy instead of BucketAccessClass.

Reply:
We have no plans to configure access policy via a CR at this time.
The BR author is not the owner of the resulting bucket -- the provisioner is the owner.

Reply:
> PV exists, because a PV can exist without PVC. Either because it was statically provisioned, or because you want to retain the state associated to it, i.e. the volume's content.

PVs exist because the lifecycle of PVs and PVCs are different. We needed a declarative way to request a volume creation/allocation, whose lifecycle was independent of the volume itself. This is why we have two resources. Not because one can exist without the other - the fact that they exist independent of one another is the effect, not the cause for their design.

Reply:
In case of BAR - BA - there is a case where BA exists after the BAR is gone. This is during deletion of credentials. The BAR goes away after calling delete on BA, which sticks around until all accessors are done using it.

Reply:
> The BAR goes away after calling delete on BA, which sticks around until all accessors are done using it.

This does not feel notable enough to justify the extra complexity of having an additional BA resource. You could instead simply keep the BAR in terminating state until it's not used anymore. I think that's how deleting a PVC works, right?

Reply:
Good point - your question made me revisit why we made the decision to have a separate resource. 

1.  The result of provisioning access credentials creates sensitive information that should not be read/mutated/deleted by non-privileged users. We achieve this by having the access information stored away in a cluster resource. 
2. The other reason for having a separate resource was to create the same style of BAR usage within pods as PVCs today - i.e. something like volumeClaimTemplates for buckets.  For such semantics, we need a separate resource because the lifecycle of the BAR itself will be tied to the Pod, whereas the BA might have to be persisted even after the Pod is gone in case the object storage vendor is down and we haven't cleaned up the resources yet.

Reply:
Ok, sounds legit.

Then I think we should relax the 1:1 mapping. I imagine that a BAR change (e.g. change of the access policy) could trigger the creation of a new BA, and deletion of the old BA as soon as

1. the new BA is ready and
2. no more Pods mount the old credentials.

(If live update of the credentials mount is implemented, then (2) is fulfilled relatively quickly. Namely, as soon as the credential mount has been updated in all Pods.)

There's a non-zero time frame during which both credentials are valid, so during that time frame both BA objects (new and old) should exist.

Reply:
@wlan0 

I just listened through the first half of the [last community meeting](https://www.youtube.com/watch?v=M3mSdI_Yj_E) (where I had no time to attend).

You proposed that, instead of using two BAs, you could have two sets of fields in one BA: blue and green. I don't think that's a good idea, and here's why:

Suppose somebody changes the BAR. The BAR-controller syncs this change to the BA and the BA-controller then initiates the switchover from blue to green. Now somebody changes the BAR again. The BAR-controller again syncs this change to the BA. But because there's already a switchover from blue to green going on, the BA-controller cannot initiate the switchover to the newest version. First, the ongoing switchover needs to finish. But what if there's a problem with one Pod that cannot be switched over?

The problem that I just described is essentially due to the limited number of slots. Two slots: blue and green.

With my proposal where there are multiple BAs, the BA-controller could always sync Pods to the newest version.